sklearn/pipeline.py

```diff
@@ -217,122 +326,144 @@ def fit_predict(self, X, y=None, **fit_params):
         X : iterable
             Training data. Must fulfill input requirements of first step of
             the pipeline.
+
         y : iterable, default=None
             Training targets. Must fulfill label requirements for all steps
             of the pipeline.
+
+        fit_params : dict of string -> object
+            Parameters passed to the `fit` method of each step, where
+            each parameter name is prefixed such that parameter ``p`` for step
+            ``s`` has key ``s__p``.
         """
-        Xt, fit_params = self._pre_transform(X, y, **fit_params)
+        Xt = X
+        for name, transform in self.steps[:-1]:
+            if transform is not None:
+                Xt = transform.transform(Xt)
         return self.steps[-1][-1].fit_predict(Xt, y, **fit_params)
 
     @if_delegate_has_method(delegate='_final_estimator')
     def predict_proba(self, X):
-        """Applies transforms to the data, and the predict_proba method of the
-        final estimator. Valid only if the final estimator implements
-        predict_proba.
+        """Apply transforms, and predict_proba of the final estimator
 
         Parameters
         ----------
-        X : iterable
-            Data to predict on. Must fulfill input requirements of first step
-            of the pipeline.
+        X : array-like, shape = [n_samples, n_features]
+            Data samples, where ``n_samples`` is the number of samples and
+            ``n_features`` is the number of features.
+
+        Returns
+        -------
+        y_proba : array-like, shape = [n_samples, n_classes]
         """
         Xt = X
         for name, transform in self.steps[:-1]:
-            Xt = transform.transform(Xt)
+            if transform is not None:
+                Xt = transform.transform(Xt)
         return self.steps[-1][-1].predict_proba(Xt)
 
     @if_delegate_has_method(delegate='_final_estimator')
     def decision_function(self, X):
-        """Applies transforms to the data, and the decision_function method of
-        the final estimator. Valid only if the final estimator implements
-        decision_function.
+        """Apply transforms, and decision_function of the final estimator
 
         Parameters
         ----------
-        X : iterable
-            Data to predict on. Must fulfill input requirements of first step
-            of the pipeline.
+        X : array-like, shape = [n_samples, n_features]
```

Comment:
This is not true, right? We don't really care about the type of X. It could be text for example.